File: elixir/leap.ex
defmodule Year do
  @doc """
  Returns whether 'year' is a leap year.

  A leap year occurs:

  on every year that is evenly divisible by 4
    except every year that is evenly divisible by 100
      unless the year is also evenly divisible by 400
  """
  @spec leap_year?(non_neg_integer) :: boolean
  def leap_year?(year) do
    s = for num <- [4, 100, 400], do: sign(rem(year, num))
    rem(Enum.sum(s), 2) == 0
  end

  @spec sign(integer) :: integer
  defp sign(n) do
    cond do
      n > 0 -> 1
      n == 0 -> 0
      n < 0 -> -1
    end
  end
end

File: zig/exer/src/leap.zig
pub fn leap_year(year: u32) bool {
    const count: u32 = sign(year % 4) + sign(year % 100) + sign(year % 400);
    return count % 2 == 0;
}

fn sign(n: u32) u32 {
    if (n > 0) return 1;
    return 0;
}
